c2rust-analyze/src/util.rs

```diff
@@ -353,3 +354,64 @@ pub fn is_null_const(constant: Constant) -> bool {
 
 pub trait PhantomLifetime<'a> {}
 impl<'a, T: ?Sized> PhantomLifetime<'a> for T {}
+
+/// Determine if `from` can be safely transmuted to `to`,
+/// which is defined as `*(from as *const To)` being a safe operation,
+/// where `from: *const From` and assuming `*from` already was safe.
+///
+/// Note that this is one-way, and is slightly different from [`core::mem::transmute`],
+/// and more similar to [`core::mem::transmute_copy`].
+///
+/// This forms a reflexive, transitive, and non-symmetric (one-way) relation, named `~` below.
+/// Formally, `A ~ B` iff `*a` and `*(a as *const B)` are safe, where `a: *const A`.
+///
+/// However, safe transmutability is difficult to check completely,
+/// so this function only checks a subset of it,
+/// with these formal rules for all types `A`, `B`:
+///
+/// * `A = B => A ~ B`
+/// * `A ~ B => *A ~ *B`
+/// * `uN ~ iN`, `iN ~ uN`, where `N` is an integer width
+/// * `[A] ~ A`
+/// * `[A; N] ~ A`, where `const N: usize`
```

Comment:
Here's an issue with zero-length arrays:
```Rust
let arr = [];
let p = &arr as *const [i32; 0];
let _ = *p;  // Well-defined: loads 0 bytes
let q = p as *const i32;
let _ = *q;  // Undefined behavior: loads 4 bytes from an object of size 0
```
There is a similar issue with zero-length slices, which is harder to address since the length is not part of the type.

Which of these rules do we actually need for lighttpd?  If we only need `*const u8 <-> *const i8` then maybe it would be best to remove the array and slice rules for now.  Otherwise, I'm not sure what we'd want to change regarding the rules and/or the definition of `A ~ B`.

Reply:
> This writeup might still be useful for reasoning about the rules, once they're fixed.
> 
> The main property we want is that if `A ~ B`, then `sizeof(A) >= sizeof(B)` and `alignof(A) >= alignof(B)`. For rules 1 and 3, this obviously holds. For rule 2, this holds as long as `A ~ B` implies that if `*B` is a fat pointer, then `*A` is also a fat pointer.
> 
> This latter property obviously holds for rules 1 and 3. For rule 4, it holds because the `*[A]` is always a fat pointer. For rule 5, the property holds because when `*A` is a fat pointer, `A` is unsized, and therefore `[A; N]` is ill-formed. For rule 2, it holds because `**A` and `**B` are always thin pointers.

I can include this in the docs if you think that's good.

Reply:
> Which of these rules do we actually need for lighttpd? If we only need `*const u8 <-> *const i8` then maybe it would be best to remove the array and slice rules for now. Otherwise, I'm not sure what we'd want to change regarding the rules and/or the definition of `A ~ B`.

It's been a while so I'm trying to remember, but I'm pretty sure we need the array rule, as that's the intermediary cast for string literals.  For arrays, I can restrict the rule to `A ~ B, N > 0 => [A; N] ~ B`, right, as `N` will be a constant I check at compile-time.

> There is a similar issue with zero-length slices, which is harder to address since the length is not part of the type.

For slices, that's a lot trickier.  Let me check if we need them first.

Reply:
I added `N > 0` as a pre-requisite to the array rule in 6b3bdd88b70103e40435cf575f4f520d8955e4ae.

For slices, I don't think we need them, so I'll remove them for now until we either need them or figure out a way to make them sound.

Reply:
I added the fat ptr proof in 7f87dc1ece162561e675d185a25e9ede4b3bf459.